pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -580,6 +558,50 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
 	}(l, vmi)
 }
 
+// startMigrationProxyServer starts a local migration proxy.
+// Right now Libvirt won't let us perform a migration using a unix socket, so
+// we have to create a local host tcp server (on port 22222) that forwards the traffic
+// to libvirt in order to trick libvirt into doing what we want.
+// This also creates a tcp server for each additional direct migration connections
+// that will be proxied to the destination pod
+func startMigrationProxyServer(virtShareDir string, vmi *v1.VirtualMachineInstance, loopbackAddress string) (stopServer func(), err error) {
+	var proxyStopListenPool []func()
+	stopServer = func() {
+		for _, f := range proxyStopListenPool {
+			f()
+		}
+	}
+	defer func() {
+		// On error, stop the servers before returning.
+		if err != nil {
+			stopServer()
+			stopServer = nil
+		}
+	}()
+
+	isBlockMigration := vmi.Status.MigrationMethod == v1.BlockMigration
+	migrationPortsRange := migrationproxy.GetMigrationPortsList(isBlockMigration)
+
+	// Create a tcp server for each direct connection proxy
+	for _, port := range migrationPortsRange {
+		key := migrationproxy.ConstructProxyKey(string(vmi.UID), port)
+		migrationProxy := migrationproxy.NewTargetProxy(loopbackAddress, port, nil, nil, migrationproxy.SourceUnixFile(virtShareDir, key))
+		if e := migrationProxy.StartListening(); e != nil {
+			return stopServer, e
+		}
+		proxyStopListenPool = append(proxyStopListenPool, migrationProxy.StopListening)
+	}
+
+	//  proxy incoming migration requests on port 22222 to the vmi's existing libvirt connection
+	libvirtConnectionProxy := migrationproxy.NewTargetProxy(loopbackAddress, LibvirtLocalConnectionPort, nil, nil, migrationproxy.SourceUnixFile(virtShareDir, string(vmi.UID)))
+	if e := libvirtConnectionProxy.StartListening(); e != nil {
+		return stopServer, e
+	}
+	proxyStopListenPool = append(proxyStopListenPool, libvirtConnectionProxy.StopListening)
```

Comment:
I have to admit for me it was very hard to understand the logic of  `proxyStopListenPool` comparing to the previous defer logic.
A method for stopping the server that invokes a pool of stopping methods which is invoked by one of the two defers, depends whether there was an error or not.
I felt like I'm solving an interview question:)
Can you please explain the added value of this. Readability is for sure not one of the added values.

Reply:
> I found no other way to implement the needed. If you have something else in mind, please let me know.
Let the caller do the clean in case of an error.
Anyway we depend on the caller doing the cleanup if there is no error.

You can leave it as is, I just find it very confusing for the "future developer".